core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java

```diff
@@ -737,6 +783,8 @@ public void close() {
       tableZooHelper.close();
     }
     singletonReservation.close();
+    this.scannerReadaheadPool.shutdownNow(); // abort all tasks, client is shutting down
+    this.cleanupThreadPool.shutdown(); // wait for shutdown tasks to execute
```

Comment:
These should be ordered before the `singletonReservation.close()` call above them. Also, I noticed the other thread pools held by this class are lazily instantiated, so could be null. Should these be lazily instantiated as well? It would save some resources when using a client for things that doesn't require them.

Reply:
I can move the `close()` calls up. Regarding lazy instantiation, what other thread pools were you referencing? The class javadoc says `Any state in this object should be available at the time of its construction.`, does that on mean configuration?

Reply:
I was referring to `thriftTransportPool`, a few lines above these. This class has evolved since I originally wrote that javadoc. While that was the original intent, I'm not sure it's entirely necessary or practical given how much state we're using in here. What is most important is that everything is available in the sense that no additional API calls are necessary to further initialize the state, that all of its functionality is ready after constructing. Lazy initialization would still be compatible with that looser definition of availability.